modules/irqchip/riscv/fdt_irqchip.c

```diff
@@ -11,6 +11,358 @@
+
+   } else if (param->param_nums == 2) {
+
+      *hwirq = param->params[0];
+      *type = param->params[1];
+
+   } else {
+
+      printk("irqchip: ERROR: wrong irq parammetet numbers\n");
+      return -EINVAL;
+   }
+
+   return 0;
+}
+
+/*
+ * Only "interrupts-extended" and "interrupts" formats have been parsed here,
+ * and the "interrupt-map" format has not been processed.
+ */
+int fdt_parse_one_hwirq(void *fdt, int node, int index,
+                        struct fdt_irq_param *irq_param)
+{
+   int len, rc;
+   int intc_node, parent_node;
+   const fdt32_t *val;
+   u32 ints_cells, intc_phandle;
+   struct fdt_phandle_args pargs;
+
+   /* 1. Try parse "interrupts-extended" fdt format */
+
+   rc = fdt_parse_phandle_with_args(fdt, node, "interrupts-extended",
+                                    "#interrupt-cells", index, &pargs);
+
+   if (!rc) {
+      irq_param->intc_node = pargs.node_offset;
+      irq_param->param_nums = pargs.args_count;
+
+      for (int i = 0; i < pargs.args_count; i++) {
+         irq_param->params[i] = pargs.args[i];
+      }
+      return 0;
+   }
+
+   /* 2. Try parse "interrupts" fdt format */
+
+   for (parent_node = node;
+        fdt_node_depth(fdt, parent_node) > 0;
+        parent_node = fdt_parent_offset(fdt, parent_node))
+   {
+      val = fdt_getprop(fdt, parent_node, "interrupt-parent", &len);
+      if (val) {
+         intc_phandle = fdt32_to_cpu(*val);
+         break;
+      }
+   }
+
+   intc_node = fdt_node_offset_by_phandle(fdt, intc_phandle);
+   if (intc_node < 0 || fdt_node_depth(fdt, parent_node) <= 0)
+      return -ENODEV;
+
+   val = fdt_getprop(fdt, intc_node, "#interrupt-cells", &len);
+   if (!val)
+      return -ENODEV;
+
+   ints_cells = fdt32_to_cpu(*val);
+   if (ints_cells > FDT_MAX_INTERRUPTS_PARAMS)
+      return -EINVAL;
+
+   val = fdt_getprop(fdt, node, "interrupts", &len);
+   if (!val)
+      return -ENODEV;
+
+   // Check if the index is within the parameter range
+   if (index * ints_cells >= len / sizeof(*val))
+      return -EINVAL;
+
+   for (u32 i = 0; i < ints_cells; i++)
+      irq_param->params[i] = fdt32_to_cpu(val[index * ints_cells + i]);
+
+   irq_param->intc_node = intc_node;
+   irq_param->param_nums = ints_cells;
+
+   return 0;
+}
+
+int irqchip_get_irq_count(void *fdt, int node)
+{
+   struct fdt_irq_param irq_param;
+   int index = 0;
+   int rc = 0;
+
+   while (rc == 0) {
+      rc = fdt_parse_one_hwirq(fdt, node, index, &irq_param);
+      index++;
+   }
+
+   return index;
+}
+
+int irqchip_alloc_irq(void *fdt, int node, int hwirq_index)
+{
+   struct irq_domain *domain, *tmp;
+   struct fdt_irq_param irq_param;
+   int irq = 0, rc;
+   int hwirq, type;
+
+   rc = fdt_parse_one_hwirq(fdt, node, hwirq_index, &irq_param);
+   if (rc)
+      return 0;
+
+   disable_preemption();
+
+   /* Find the corresponding irq domain */
+   list_for_each(domain, tmp, &irq_domains, node) {
+
+      if (domain->fdt_node == irq_param.intc_node) {
```

Comment:
Can you do:
```
if (domain->fdt_node != irq_param.intc_node)
   continue;
```

Instead? That will one extra level of indentation will go away.

Reply:
In alternative, you can do the `if` in the positive case, but extract the body of that `if` to a separate function, so that it's easier to read.

Reply:
That makes sense！